crates/editor/src/element.rs

```diff
@@ -1558,6 +1589,258 @@ impl EditorElement {
+            })
+            .sorted_by_key(|diagnostic| {
+                (
+                    diagnostic.diagnostic.severity,
+                    std::cmp::Reverse(diagnostic.diagnostic.is_primary),
+                    diagnostic.range.start.row,
+                    diagnostic.range.start.column,
+                )
+            });
+
+        let max_entries = 5; // ToDo: Configuration
+        for diagnostic in diagnostics {
+            let diag_point = MultiBufferPoint::new(diagnostic.range.start.row, 0);
+            let display_point = snapshot.point_to_display_point(diag_point, Bias::Right);
+
+            if !(start_row..end_row).contains(&display_point.row()) {
+                continue;
+            }
+            let line_ix = DisplayRow(display_point.row().minus(start_row));
+
+            if let Some(ref mut elem) = inline_diagnostics.get_mut(&line_ix) {
+                if elem.entries.len() >= max_entries {
+                    elem.exceeded_max = true;
+                    continue;
+                }
+
+                let entry = DiagnosticDisplay::new(
+                    diagnostic.diagnostic.severity,
+                    diagnostic.diagnostic.group_id,
+                    cx.new_view(|cx| {
+                        DiagnosticTooltip::new(diagnostic, hitbox, line_height, em_width, cx)
+                    }),
+                );
+                elem.entries.push(entry);
+                continue;
+            }
+
+            let start_y = content_origin.y
+                + line_height
+                    * (display_point.row().0 as f32 - scroll_pixel_position.y / line_height);
+
+            let start_x = {
+                const INLINE_BLAME_PADDING_EM_WIDTHS: f32 = 6.;
+
+                let crease_trailer_layout = crease_trailers[line_ix.0 as usize].as_ref();
+                let line_layout = &line_layouts[line_ix.0 as usize];
+
+                let line_end = if let Some(crease_trailer) = crease_trailer_layout {
+                    crease_trailer.bounds.right()
+                } else {
+                    content_origin.x - scroll_pixel_position.x + line_layout.width
+                };
+                let padded_line_end = line_end + em_width * INLINE_BLAME_PADDING_EM_WIDTHS;
+
+                let min_column = ProjectSettings::get_global(cx)
+                    .diagnostics
+                    .inline()
+                    .min_column();
+                let min_column_in_pixels = self.column_pixels(min_column as usize, cx);
+                let min_start = content_origin.x - scroll_pixel_position.x + min_column_in_pixels;
+
+                cmp::max(padded_line_end, min_start)
+            };
+
+            let entry = DiagnosticDisplay::new(
+                diagnostic.diagnostic.severity,
+                diagnostic.diagnostic.group_id,
+                cx.new_view(|cx| {
+                    DiagnosticTooltip::new(diagnostic.clone(), hitbox, line_height, em_width, cx)
```

Comment:
NIT: looks like `diagnostic.clone()` could be avoided if we preallocate `message` and `severity` before?
Overall, a good practice to keep less non-pointer clones, esp. around the rendering code.

Reply:
This is also gone so no longer applicable.